File: 567. Permutation in String/solution.go
package main

func isSameMap(map1, map2 map[string]int) bool {
	if len(map1) != len(map2) {
		return false
	}
	for k, v := range map1 {
		if v != map2[k] {
			return false
		}
	}
	return true
}

func checkInclusion(s1 string, s2 string) bool {
    if len(s2) < len(s1) {
		return false
	}
	// put s1 chars is s1Map
	s1Map := make(map[string]int)
	for _, char := range s1 {
		s1Map[string(char)]++
	}
	// init s2Map
	s2Map := make(map[string]int)
	for i := 0; i < len(s1); i++ {
		s2Map[string(s2[i])]++
	}
	// compare the first window
	if isSameMap(s1Map, s2Map) {
		return true
	}
	// make a for loop start from index len(s1) to len(s2)
	// in each loop delete left most element and add one more element to the right
	for i := len(s1); i < len(s2); i++ {
		left := i-len(s1)
		right := i
		s2Map[string(s2[left])]--
		if s2Map[string(s2[left])] == 0 {
			delete(s2Map, string(s2[left]))
		}
		s2Map[string(s2[right])]++
		if isSameMap(s1Map, s2Map) {
			return true
		}
	}
	return false
}

File: 567. Permutation in String/solution.py
"""
Given two strings s1 and s2, return true if s2 contains a permutation of s1, or false otherwise.

In other words, return true if one of s1's permutations is the substring of s2.
"""

"""
Note:
1. s2 len should be greater than that of s1
"""

from collections import defaultdict


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # if len(s2) < len(s1), return False
        if len(s2) < len(s1):
            return False
        # init a n variable represent the length of s1
        n = len(s1)
        # init a s1 map and s2 map to store char count in window
        s1_map = defaultdict(int)
        s2_map = defaultdict(int)
        for c in s1:
            s1_map[c]+=1
        for i in range(n):
            s2_map[s2[i]]+=1
        # implement a is_same_map method to compare whether 2 map are equal
        def is_same_map(map1: dict, map2: dict) -> bool:
            if len(map1) != len(map2):
                return False
            for k, v in map1.items():
                if v != map2[k]:
                    return False
            return True
        # compare 2 maps once
        if is_same_map(s1_map, s2_map):
            return True
        # iterate s2 from n
        for i in range(n, len(s2)):
            # adjust the s2_map by adding s2[i] and delete s2[i-n-1]
            s2_map[s2[i]]+=1
            s2_map[s2[i-n]]-=1
            if s2_map[s2[i-n]] == 0:
                del s2_map[s2[i-n]]
            # compare 2 map, if True then return
            if is_same_map(s1_map, s2_map):
                return True
        # return False as no permutation
        return False